File: Questions/Array & Hashing/groupAnagrams.py
strs1 = ["eat","tea","tan","ate","nat","bat"]
# output = [["bat"],["nat","tan"],["ate","eat","tea"]]
strs2 = [""]
# output = [[""]]
strs3 = ["a"]
# output = [["a"]]

def groupAnagrams(strs): # 111ms
    groups = {}
    for word in strs: # O(n)
        d = ''.join(sorted(word)) # O(nlogn)
        if d in groups: # will always compare with sorted value
            groups[d].append(word)
        else:
            groups[d] = [word]
    anagrams = []
    for key, value in groups.items():
        anagrams.append(value)
    return anagrams

print(groupAnagrams(strs1))

File: javascript/array/groupAnagrams.js
// group anagrams
// i/p = ["eat","tea","tan","ate","nat","bat"]
// o/p = [["bat"],["nat","tan"],["ate","eat","tea"]]

function groupAnagrams(strs) {
    // TC: O(n)
    // SC: O(n)
    let anagrams = {};
    for (let i=0; i<strs.length; i++) {
        const word = strs[i];
        if (anagrams.hasOwnProperty(word.split("").sort().join(""))) {
            anagrams[word.split("").sort().join("")].push(word);
        } else {
            anagrams[word.split("").sort().join("")] = [word];
        }
    }
    const group_anagrams = [];
    for (key in anagrams) {
        group_anagrams.push(anagrams[key]);
    }
    return group_anagrams;
}

console.log(groupAnagrams(["eat","tea","tan","ate","nat","bat"]));